### PR TITLE
Use protobuf to serialize trie node

### PR DESCRIPTION
**Pull request by dustinxie**
1. Testing shows for Branch node, serialization size reduced 75%, for Leaf node the reduction is 50%
2. The running time of perform 2k Insertion/Deletion reduced 50%

### CODE REVIEW

File: trie/branch.go [hunk 1]
```diff
@@ -154,29 +153,42 @@ func (b *branch) hash() hash.Hash32B {
 	return blake2b.Sum256(stream)
 }
 
+func (b *branch) toProto() *iproto.NodePb {
```
Comment (zjshen14) [inline trie/branch.go]: I prefer call it `ToProto` as we have widely use throughout the code. Later on, we could extract interface from this.

File: trie/branch.go [hunk 2]
```diff
@@ -154,29 +153,42 @@ func (b *branch) hash() hash.Hash32B {
...
-	}
-	// first byte denotes the type of patricia: 1-branch, 0-ext/leaf
-	return append([]byte{BRANCH}, stream.Bytes()...), nil
+	return proto.Marshal(b.toProto())
 }
 
-// deserialize to branch
-func (b *branch) deserialize(stream []byte) error {
-	// reset variable
+func (b *branch) fromProto(pbBranch *iproto.BranchPb) {
```
Comment (zjshen14) [inline trie/branch.go]: Ditto

File: trie/branch.go [hunk 3]
```diff
@@ -154,29 +153,42 @@ func (b *branch) hash() hash.Hash32B {
...
+	}
 }
 
-func (b *branch) print() {
-	for i := 0; i < RADIX; i++ {
-		if len(b.Path[i]) > 0 {
-			logger.Info().Int("k", i).Hex("v", b.Path[i]).Msg("branch")
-		}
+// deserialize to branch
+func (b *branch) deserialize(stream []byte) error {
```
Comment (zjshen14) [inline trie/branch.go]: Same as From/ToProto

File: trie/trie.go [hunk 4]
```diff
@@ -73,17 +73,34 @@ type (
...
 	}
-	return newTrieSharedBatch(kvStore, batch, name, root), nil
+	t := &trie{
+		cb:        batch,
+		dao:       kvStore,
+		rootHash:  root,
+		bucket:    name,
+		numEntry:  1,
+		numBranch: 1}
+	t.lifecycle.Add(kvStore)
```
Comment (zjshen14) [inline trie/trie.go]: I doubt this is a safe operation:

1. We don't want to start/stop kv store upon each batch/block cycle
2. We don't know kvStore is started/stopped here. It's better for the creator to start/stop it.

Comment (zjshen14) [reply trie/trie.go]: In fact, I think it's never been used, because working set doesn't use lifecycle

File: trie/patricia_test.go [hunk 5]
```diff
@@ -51,48 +51,47 @@ func TestPatricia(t *testing.T) {
...
 	assert.NotNil(stream)
-	assert.Equal(byte(1), stream[0])
 	b1 := branch{}
 	err = b1.deserialize(stream)
 	assert.Nil(err)
 	assert.Equal(0, bytes.Compare(root, b1.Path[0]))
 	assert.Equal(0, bytes.Compare(hash1, b1.Path[2]))
 	assert.Equal(0, bytes.Compare(hash2, b1.Path[11]))
-	assert.Equal(430, len(stream))
+	assert.Equal(114, len(stream))
```
Comment (dustinxie) [inline trie/patricia_test.go]: size reduce to 1/4 of using Gob

File: trie/patricia_test.go [hunk 6]
```diff
@@ -51,48 +51,47 @@ func TestPatricia(t *testing.T) {
...
 	err = e1.deserialize(stream)
 	assert.Nil(err)
+	assert.Equal(0, e1.Ext)
 	assert.Equal(hash1, e1.Value)
 	assert.Equal(byte(2), e1.Path[0])
 	assert.Equal(byte(3), e1.Path[1])
 	assert.Equal(byte(5), e1.Path[2])
 	assert.Equal(byte(7), e1.Path[3])
-	assert.Equal(91, len(stream))
+	assert.Equal(42, len(stream))
```
Comment (dustinxie) [inline trie/patricia_test.go]: size reduce to 1/2 of using Gob

File: trie/trie_test.go [hunk 7]
```diff
@@ -482,11 +489,11 @@ func Test2kEntries(t *testing.T) {
 	require.Nil(tr.Start(context.Background()))
 	root := EmptyRoot
 	seed := time.Now().Nanosecond()
-	// insert 2k entries
+	// insert 4k entries
```
Comment (dustinxie) [inline trie/trie_test.go]: in the same amount of time, can test 4k entries. meaning protobuf is 2x faster than Gob

File: trie/trie.go (hunk 4 above)
Comment (dustinxie) [reply trie/trie.go]: yes, this can be removed, since the kvStore here must have been already added

File: trie/branch.go (hunk 1 above)
Comment (dustinxie) [reply trie/branch.go]: after discussion, this is an internal data struct, will keep as-is, can change to public if we need to expose interface from this

File: trie/branch.go [hunk 8]
```diff
@@ -7,15 +7,14 @@
 package trie
 
 import (
-	"bytes"
-	"encoding/gob"
-
+	"github.com/golang/protobuf/proto"
```
Comment (zjshen14) [inline trie/branch.go]: wrong import group

**Pull request merged**